pkg/util/interrupt.go

```diff
@@ -0,0 +1,95 @@
+/*
+Derived from:
+
+https://github.com/kubernetes/kubernetes/blob/master/pkg/util/interrupt/interrupt.go
+
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+)
+
+// childSignals are the allowed signals that can be sent to children in Unix variant OS's
+var childSignals = []os.Signal{syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT}
+
+type InterruptHandler struct {
+	notify []func()
+	final  func(os.Signal)
+	once   sync.Once
+}
+
+// New creates a new Handler.
+// The first argument will be called only if a terminating signal is received, it will be the last
+// function called.  It is this function's responsibility to call os.Exit() if exiting is desired.
+// If the first argument is nil, this Handler will call os.Exit when the signal is received.
+// The second+ arguments are functions that will be called when the signal is received, or after
+// the argument to Run() completes.
+func NewInterruptHandler(final func(os.Signal), notify ...func()) *InterruptHandler {
+	return &InterruptHandler{
+		final:  final,
+		notify: notify,
+	}
+}
+
+// Close invokes the notify functions of the Handler
+// when no signal was received and Run() completed.
+func (h *InterruptHandler) Close() {
+	h.once.Do(func() {
+		for _, fn := range h.notify {
+			fn()
+		}
+	})
+}
+
+// Signal invokes the notify functions of the Handler
+// when a signal is received.  It also invokes final,
+// if any, otherwise it invokes os.Exit.
+func (h *InterruptHandler) Signal(s os.Signal) {
+	h.once.Do(func() {
+		for _, fn := range h.notify {
+			fn()
+		}
+		if h.final == nil {
+			os.Exit(0)
+		}
+		h.final(s)
+	})
+}
+
+// Run takes a single function argument which will be immediately invoked.
+// If a signal arrives while the function is running, it will be handled,
+// interrupting the function execution and executing any final and notify
+// functions immediately.  The original function may or may not complete
+// depending on whether the final argument invoking os.Exit.
+func (h *InterruptHandler) Run(fn func() error) error {
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, childSignals...)
+	defer signal.Stop(ch)
```

Comment:
team work working for us again!

Rodolfo Carvalho

OpenShift Developer Experience

On Fri, Apr 1, 2016 at 3:19 PM, Ben Parees <notifications@github.com> wrote:

> In pkg/util/interrupt.go
> <https://github.com/openshift/source-to-image/pull/452#discussion_r58203776>
> :
>
> > +		if h.final == nil {
> > +			os.Exit(0)
> > +		}
> > +		h.final(s)
> > +	})
> > +}
> > +
> > +// Run takes a single function argument which will be immediately invoked.
> > +// If a signal arrives while the function is running, it will be handled,
> > +// interrupting the function execution and executing any final and notify
> > +// functions immediately.  The original function may or may not complete
> > +// depending on whether the final argument invoking os.Exit.
> > +func (h *InterruptHandler) Run(fn func() error) error {
> > +	ch := make(chan os.Signal, 1)
> > +	signal.Notify(ch, childSignals...)
> > +	defer signal.Stop(ch)
>
> sigh. and it was in my head as i was about to do it "don't forget to fix
> the leak" and then i totally forgot. thanks :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/source-to-image/pull/452/files/7c3aeb752fb82b3002800315666fb9f2c039e56c#r58203776>
>